exercises/react/src/test/kotlin/ReactTest.kt

```diff
@@ -0,0 +1,208 @@
+        input.value = 4
+        assertEquals(listOf(222), vals)
+    }
+
+    @Test
+    fun callbacksCanBeAddedAndRemoved() {
+        val reactor = Reactor<Int>()
+        val input = reactor.InputCell(11)
+        val output = reactor.ComputeCell(input) { it + 1 }
+
+        val vals1 = mutableListOf<Int>()
+        val sub1 = output.addCallback { vals1.add(it) }
+        val vals2 = mutableListOf<Int>()
+        output.addCallback { vals2.add(it) }
+
+        input.value = 31
+        sub1.cancel()
+
+        val vals3 = mutableListOf<Int>()
+        output.addCallback { vals3.add(it) }
+
+        input.value = 41
+
+        assertEquals(listOf(32), vals1)
+        assertEquals(listOf(32, 42), vals2)
+        assertEquals(listOf(42), vals3)
+    }
+
+    @Test
+    fun removingACallbackMultipleTimesDoesntInterfereWithOtherCallbacks() {
+        val reactor = Reactor<Int>()
+        val input = reactor.InputCell(1)
+        val output = reactor.ComputeCell(input) { it + 1 }
+
+        val vals1 = mutableListOf<Int>()
+        val sub1 = output.addCallback { vals1.add(it) }
+        val vals2 = mutableListOf<Int>()
+        output.addCallback { vals2.add(it) }
+
+        for (i in 1..10) {
+            sub1.cancel()
+        }
+
+        input.value = 2
+        assertEquals(listOf<Int>(), vals1)
+        assertEquals(listOf(3), vals2)
+    }
+
+    @Test
+    fun callbacksShouldOnlyBeCalledOnceEvenIfMultipleDependenciesChange() {
+        val reactor = Reactor<Int>()
+        val input = reactor.InputCell(1)
+        val plusOne = reactor.ComputeCell(input) { it + 1 }
+        val minusOne1 = reactor.ComputeCell(input) { it - 1 }
+        val minusOne2 = reactor.ComputeCell(minusOne1) { it - 1 }
+        val output = reactor.ComputeCell(plusOne, minusOne2) { x, y -> x * y }
+
+        val vals = mutableListOf<Int>()
+        output.addCallback { vals.add(it) }
+
+        input.value = 4
+        assertEquals(listOf(10), vals)
+    }
+
+    @Test
+    fun callbacksShouldNotBeCalledIfDependenciesChangeButOutputValueDoesntChange() {
+        val reactor = Reactor<Int>()
+        val input = reactor.InputCell(1)
+        val plusOne = reactor.ComputeCell(input) { it + 1 }
+        val minusOne = reactor.ComputeCell(input) { it - 1 }
+        val alwaysTwo = reactor.ComputeCell(plusOne, minusOne) { x, y -> x - y }
+
+        val vals = mutableListOf<Int>()
+        alwaysTwo.addCallback { vals.add(it) }
+
+        for (i in 1..10) {
+            input.value = i
+        }
+
+        assertEquals(listOf<Int>(), vals)
+    }
+}
+
+@RunWith(Parameterized::class)
+// This is a digital logic circuit called an adder:
+// https://en.wikipedia.org/wiki/Adder_(electronics)
+class ReactAdderTest(val aval: Boolean, val bval: Boolean, val cin: Boolean, val expectedCout: Boolean, val expectedSum: Boolean) {
+    companion object {
+        @JvmStatic
+        @Parameterized.Parameters(name = "{index}: {0} + {1} + {2} = {3} {4}")
+        fun data() = listOf(
```

Comment:
All of these parameters make it a bit difficult to read, perhaps something like this would be an improvement:

```
diff --git a/exercises/react/src/test/kotlin/ReactTest.kt b/exercises/react/src/test/kotlin/ReactTest.kt
index 92cc9e0..a514b6b 100644
--- a/exercises/react/src/test/kotlin/ReactTest.kt
+++ b/exercises/react/src/test/kotlin/ReactTest.kt
@@ -169,28 +169,32 @@ class ReactTest {
 @RunWith(Parameterized::class)
 // This is a digital logic circuit called an adder:
 // https://en.wikipedia.org/wiki/Adder_(electronics)
-class ReactAdderTest(val aval: Boolean, val bval: Boolean, val cin: Boolean, val expectedCout: Boolean, val expectedSum: Boolean) {
+class ReactAdderTest(val input: Input, val expected: Expected) {
+
     companion object {
+        data class Input(val a: Boolean, val b: Boolean, val carryIn: Boolean)
+        data class Expected(val carryOut: Boolean, val sum: Boolean)
+
         @JvmStatic
-        @Parameterized.Parameters(name = "{index}: {0} + {1} + {2} = {3} {4}")
+        @Parameterized.Parameters(name = "{index}: {0} = {1}")
         fun data() = listOf(
-                arrayOf(false, false, false, false, false),
-                arrayOf(false, false, true, false, true),
-                arrayOf(false, true, false, false, true),
-                arrayOf(false, true, true, true, false),
-                arrayOf(true, false, false, false, true),
-                arrayOf(true, false, true, true, false),
-                arrayOf(true, true, false, true, false),
-                arrayOf(true, true, true, true, true)
+                arrayOf(Input(a=false, b=false, carryIn=false), Expected(carryOut=false, sum=false)),
+                arrayOf(Input(a=false, b=false, carryIn=true),  Expected(carryOut=false, sum=true)),
+                arrayOf(Input(a=false, b=true,  carryIn=false), Expected(carryOut=false, sum=true)),
+                arrayOf(Input(a=false, b=true,  carryIn=true),  Expected(carryOut=true,  sum=false)),
+                arrayOf(Input(a=true,  b=false, carryIn=false), Expected(carryOut=false, sum=true)),
+                arrayOf(Input(a=true,  b=false, carryIn=true),  Expected(carryOut=true,  sum=false)),
+                arrayOf(Input(a=true,  b=true,  carryIn=false), Expected(carryOut=true,  sum=false)),
+                arrayOf(Input(a=true,  b=true,  carryIn=true),  Expected(carryOut=true,  sum=true))
         )
     }
 
     @Test
     fun test() {
         val reactor = Reactor<Boolean>()
-        val a = reactor.InputCell(aval)
-        val b = reactor.InputCell(bval)
-        val carryIn = reactor.InputCell(cin)
+        val a = reactor.InputCell(input.a)
+        val b = reactor.InputCell(input.b)
+        val carryIn = reactor.InputCell(input.carryIn)
 
         val aXorB = reactor.ComputeCell(a, b) { a, b -> a.xor(b) }
         val sum = reactor.ComputeCell(aXorB, carryIn) { axorb, cin -> axorb.xor(cin) }
@@ -199,10 +203,6 @@ class ReactAdderTest(val aval: Boolean, val bval: Boolean, val cin: Boolean, val
         val aAndB = reactor.ComputeCell(a, b) { a, b -> a && b }
         val carryOut = reactor.ComputeCell(aXorBAndCin, aAndB) { a, b -> a || b }
 
-        // Test them both at once so if they fail we get to see both values.
-        assertEquals(
-                listOf(expectedSum, expectedCout),
-                listOf(sum.value, carryOut.value)
-        )
+        assertEquals(expected, Expected(sum=sum.value, carryOut=carryOut.value))
     }
 }
```